### PR TITLE
Initial version with support for parameter annotation on objective fn.

### PR DESCRIPTION
**Pull request by brunodoamaral**
Proposed implementation of https://github.com/hyperopt/hyperopt/issues/769.

### CODE REVIEW

**Review by marload — APPROVED** [review]
Great! LGTM

File: hyperopt/exceptions.py
Comment (marctorsoc) [inline hyperopt/exceptions.py:41]:
```suggestion
class InvalidAnnotatedParameter(ValueError):
```
?

File: hyperopt/exceptions.py
Comment (marctorsoc) [inline hyperopt/exceptions.py:45]: cool way to init, didn't see this before 💯

File: hyperopt/fmin.py
Comment (marctorsoc) [inline hyperopt/fmin.py:103]: the typical way to abbreviate this is args and kwargs, but not kargs

File: hyperopt/fmin.py
Comment (marctorsoc) [inline hyperopt/fmin.py:531]: given it's a parameter, what if we call it `param` ?

Comment (marctorsoc) [reply hyperopt/fmin.py:531]: Either if we make "auto" the default value for `space` or not, I would give more info in this msg error, e.g:

```
When using `space="auto"`, please annotate the objective function arguments with a `pyll.base.Apply` subclass. See example in `fmin` docstring
```

File: hyperopt/tests/test_fmin.py
Comment (marctorsoc) [inline hyperopt/tests/test_fmin.py:387]: can we do this in pytest? we're trying to move away (very slowly) from nose

**Review by marctorsoc — COMMENTED** [review]
Really cool idea, great job! it's clean and very powerful change. Some general comments:

* I would document it a bit more. Not sure if we can add an example to the docs to make it clearer. See the great work by @marload here https://github.com/hyperopt/hyperopt/tree/master/docs/templates/tutorials
* Do we want space="auto" to be the default?
* What do you think about "annotated" rather than "auto"? the latter is a bit too generic IMO

File: hyperopt/fmin.py
Comment (brunodoamaral) [reply hyperopt/fmin.py:103]: Indeed... I'll update the code

File: hyperopt/fmin.py
Comment (brunodoamaral) [reply hyperopt/fmin.py:531]: Great suggestion

File: hyperopt/tests/test_fmin.py
Comment (brunodoamaral) [reply hyperopt/tests/test_fmin.py:387]: I'll change this as requested.

**Review by brunodoamaral — COMMENTED** [review]
Resolved in commit https://github.com/hyperopt/hyperopt/pull/777/commits/8a48244c33704c828980a77fe2896dfd5cf1e2c4

**Comment by brunodoamaral** [comment]
@marctorsoc regarding your considerations:

 - I'll append to the tutorial notebook a few cells doing the same optimization, but with annotated space.
 - _Do we want space="auto" to be the default?_ - I would suggest to do it after a few months of usage. If the community thinks it is a good idea, we could leave it as default.
 - _What do you think about "annotated" rather than "auto"? the latter is a bit too generic IMO_ - indeed, "auto" sounds like we doesn't need to provide the space information.

At last, but not least, I'd like to provide an argument against my onw implementation: one may think that [type hint](https://docs.python.org/3/library/typing.html) is being misused in this approach. IMO it was designed to help the linter during development... not to incorporate information during runtime. Anyway, I still think it is a good idea to declare parameters and space in the same location.

**Comment by marctorsoc** [comment]
> @marctorsoc regarding your considerations:
> 
> * I'll append to the tutorial notebook a few cells doing the same optimization, but with annotated space.

👍  let me know when it's finished and ready to review

> * _Do we want space="auto" to be the default?_ - I would suggest to do it after a few months of usage. If the community thinks it is a good idea, we could leave it as default.

fair enough

> * _What do you think about "annotated" rather than "auto"? the latter is a bit too generic IMO_ - indeed, "auto" sounds like we doesn't need to provide the space information.

great, let's go with "annotated" then. I had the same feeling, that if it's auto you don't have to do anything. And that's misleading
 
> At last, but not least, I'd like to provide an argument against my onw implementation: one may think that [type hint](https://docs.python.org/3/library/typing.html) is being misused in this approach. IMO it was designed to help the linter during development... not to incorporate information during runtime. Anyway, I still think it is a good idea to declare parameters and space in the same location.

the 3 maintainers of this repo approved either here or in the issue you opened, so you have the greenlight :)

**Comment by marctorsoc** [comment]
> > @marctorsoc regarding your considerations:
> > 
> > * I'll append to the tutorial notebook a few cells doing the same optimization, but with annotated space.
> 
> 👍 let me know when it's finished and ready to review
> 
> > * _Do we want space="auto" to be the default?_ - I would suggest to do it after a few months of usage. If the community thinks it is a good idea, we could leave it as default.
> 
> fair enough
> 
> > * _What do you think about "annotated" rather than "auto"? the latter is a bit too generic IMO_ - indeed, "auto" sounds like we doesn't need to provide the space information.
> 
> great, let's go with "annotated" then. I had the same feeling, that if it's auto you don't have to do anything. And that's misleading
> 
> > At last, but not least, I'd like to provide an argument against my onw implementation: one may think that [type hint](https://docs.python.org/3/library/typing.html) is being misused in this approach. IMO it was designed to help the linter during development... not to incorporate information during runtime. Anyway, I still think it is a good idea to declare parameters and space in the same location.
> 
> the 3 maintainers of this repo approved either here or in the issue you opened, so you have the greenlight :)

is this ready to review now? @brunodoamaral

**Comment by brunodoamaral** [comment]
Sorry for slow response. It went out of my radar these days.
Here is my proposal of first tutorial update: https://github.com/brunodoamaral/hyperopt/blob/master/tutorial/01.BasicTutorial.ipynb

**Comment by maxpumperla** [comment]
@brunodoamaral I love what you did there, great improvement. I think your test still says "auto", which is probably why travis fails. Can you have another closer look and let me know when you're good to go with this?

**Comment by marctorsoc** [comment]
thanks for your changes @brunodoamaral . It's a bit strange that we have /tutorials and then /docs/templates/tutorials. We should reorganise this at some point... 

anyways, there seems not possible (or at least doesn't work for me) to render these ipynb files in github and also revieweing ipynb files is not easy. Do you mind using jupytext? we can keep both the .py and .ipynb, or just the .py, I'm not strongly opinionated on this

**Comment by brunodoamaral** [comment]
@marctorsoc I had no problems rendering the .ipynb on github (but sometimes I have to click on "reload" if failed). I don't think duplicating the code as a .py file will help... as the .ipynb has a lot more information, such as charts and rich text.

**Comment by marctorsoc** [comment]
> @marctorsoc I had no problems rendering the .ipynb on github (but sometimes I have to click on "reload" if failed). I don't think duplicating the code as a .py file will help... as the .ipynb has a lot more information, such as charts and rich text.

I'm not opposed to keep .ipynb files. My point is that in PRs we shouldn't review .ipynb but .py. If you want to keep them both, I'm happy with it and it's really easy using jupytext. 

In other OSS projects like sklearn, they keep only the .py, and they have a build process to generate the .ipynb automatically, see https://github.com/scikit-learn/scikit-learn/blob/8ea176ae0ca535cdbfad7413322bbc3e54979e4d/.binder/postBuild#L27. What I propose makes your life easier, I think, as you can develop your notebook in jupyter, and push both files

**Comment by marload** [comment]
I want to merge this PR. Is there any progress?
cc. @brunodoamaral

**Comment by marctorsoc** [comment]
it's fine. Let's merge and do the jupytext stuff in some other iteration

**Comment by brunodoamaral** [comment]
@marctorsoc thanks for the help. I was not familiar with jupytext and actually didn't understand how to procedure with your request. I will watch the changes made by the team to get a better picture on how to use it.

**Pull request merged**
